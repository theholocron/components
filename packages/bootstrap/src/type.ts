import * as React from "react";

export interface WithChildren {
	children?: React.ReactNode;
}
