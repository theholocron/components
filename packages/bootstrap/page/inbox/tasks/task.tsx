import { ActionIcon, Checkbox, Group, TextInput } from "@mantine/core";
import { IconStar, IconTrash } from "@tabler/icons-react";

export interface TaskProps {
	task: {
		id: string;
		title: string;
		state: "TASK_INBOX" | "TASK_PINNED" | "TASK_ARCHIVED";
	};
	onArchiveTask: (archive: "ARCHIVE_TASK", id: string) => void;
	onTogglePinTask: (
		state: "TASK_INBOX" | "TASK_PINNED" | "TASK_ARCHIVED",
		id: string
	) => void;
	onEditTitle: (title: string, id: string) => void;
	onDeleteTask: (id: string) => void;
}

export function Task(props: TaskProps) {
	const {
		task: { id, title, state },
		onArchiveTask,
		onTogglePinTask,
		onEditTitle,
		onDeleteTask,
	} = props;

	return (
		<Group
			align="center"
			justify="space-between"
			wrap="nowrap"
			gap="xl"
			role="listitem"
			aria-label={`task-${id}`}
		>
			<Checkbox
				type="checkbox"
				aria-label={`archiveTask-${id}`}
				onChange={() => onArchiveTask("ARCHIVE_TASK", id)}
				name="checked"
				id={`archiveTask-${id}`}
				checked={state === "TASK_ARCHIVED"}
				pl={10}
			/>

			<TextInput
				variant="unstyled"
				type="text"
				value={title}
				name="title"
				placeholder="Input title"
				style={{
					textOverflow: "ellipsis",
					width: "100%",
				}}
				onChange={(e) => onEditTitle(e.target.value, id)}
			/>
			<ActionIcon
				variant="transparent"
				onClick={() => onDeleteTask(id)}
				aria-label="delete"
			>
				<IconTrash
					style={{ width: "70%", height: "70%" }}
					stroke={1.5}
				/>
			</ActionIcon>
			{state !== "TASK_ARCHIVED" && (
				<ActionIcon
					variant="transparent"
					onClick={() => onTogglePinTask(state, id)}
					id={`pinTask-${id}`}
					aria-label={state === "TASK_PINNED" ? "unpin" : "pin"}
					key={`pinTask-${id}`}
					mr={10}
				>
					<IconStar
						style={{ width: "70%", height: "70%" }}
						stroke={1.5}
					/>
				</ActionIcon>
			)}
		</Group>
	);
}
