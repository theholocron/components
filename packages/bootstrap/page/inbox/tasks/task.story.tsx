import type { Meta, StoryObj } from "@storybook/react";
import { Task } from "./task";

const meta: Meta<typeof Task> = {
	component: Task,
	title: "Task",
	argTypes: {
		onArchiveTask: { action: "onArchiveTask" },
		onTogglePinTask: { action: "onTogglePinTask" },
		onEditTitle: { action: "onEditTitle" },
		onDeleteTask: { action: "onDeleteTask" },
	},
};
export default meta;
type Story = StoryObj<typeof Task>;

export const Default: Story = {
	render: (args) => (
		<ul>
			<Task {...args} />
		</ul>
	),
	args: {
		task: {
			id: "1",
			title: "Buy milk",
			state: "TASK_INBOX",
		},
	},
};

export const Pinned: Story = {
	render: (args) => (
		<ul>
			<Task {...args} />
		</ul>
	),
	args: {
		task: {
			id: "2",
			title: "QA dropdown",
			state: "TASK_PINNED",
		},
	},
};

export const Archived: Story = {
	render: (args) => (
		<ul>
			<Task {...args} />
		</ul>
	),
	args: {
		task: {
			id: "3",
			title: "Write schema for account menu",
			state: "TASK_ARCHIVED",
		},
	},
};

const longTitleString: string = `This task's name is absurdly large. In fact, I think if I keep going I might end up with content overflow. What will happen? The star that represents a pinned task could have text overlapping. The text could cut-off abruptly when it reaches the star. I hope not!`;

export const LongTitle: Story = {
	args: {
		task: {
			id: "4",
			title: longTitleString,
			state: "TASK_INBOX",
		},
	},
	render: (args) => (
		<ul>
			<Task {...args} />
		</ul>
	),
};
