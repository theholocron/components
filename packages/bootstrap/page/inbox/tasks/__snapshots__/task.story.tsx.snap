// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Archived smoke-test 1`] = `
<ul>
  <div class="list-item TASK_ARCHIVED"
       role="listitem"
       aria-label="task-3"
  >
    <label for="checked"
           aria-label="archiveTask-3"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-3"
             checked
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-3"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Write schema for account menu"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Write schema for account menu"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
  </div>
</ul>
`;

exports[`Task Default smoke-test 1`] = `
<ul>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-1"
  >
    <label for="checked"
           aria-label="archiveTask-1"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-1"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-1"
      >
      </span>
    </label>
    <label for="title"
           aria-label="Buy milk"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="Buy milk"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-1"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
</ul>
`;

exports[`Task LongTitle smoke-test 1`] = `
<ul>
  <div class="list-item TASK_INBOX"
       role="listitem"
       aria-label="task-4"
  >
    <label for="checked"
           aria-label="archiveTask-4"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-4"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-4"
      >
      </span>
    </label>
    <label for="title"
           aria-label="This task's name is absurdly large. In fact, I think if I keep going I might end up with content overflow. What will happen? The star that represents a pinned task could have text overlapping. The text could cut-off abruptly when it reaches the star. I hope not!"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="This task's name is absurdly large. In fact, I think if I keep going I might end up with content overflow. What will happen? The star that represents a pinned task could have text overlapping. The text could cut-off abruptly when it reaches the star. I hope not!"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-4"
            aria-label="pin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
</ul>
`;

exports[`Task Pinned smoke-test 1`] = `
<ul>
  <div class="list-item TASK_PINNED"
       role="listitem"
       aria-label="task-2"
  >
    <label for="checked"
           aria-label="archiveTask-2"
           class="checkbox"
    >
      <input type="checkbox"
             disabled
             name="checked"
             id="archiveTask-2"
      >
      <span class="checkbox-custom"
            role="button"
            aria-label="archiveButton-2"
      >
      </span>
    </label>
    <label for="title"
           aria-label="QA dropdown"
           class="title"
    >
      <input type="text"
             name="title"
             placeholder="Input title"
             value="QA dropdown"
             style="text-overflow: ellipsis;"
      >
    </label>
    <button aria-label="delete"
            class="delete-button"
    >
      <span class="icon-trash">
      </span>
    </button>
    <button class="pin-button"
            id="pinTask-2"
            aria-label="unpin"
    >
      <span class="icon-star">
      </span>
    </button>
  </div>
</ul>
`;
